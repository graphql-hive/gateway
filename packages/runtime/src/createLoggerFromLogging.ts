import { Logger, LogLevel } from '@graphql-hive/logger';

export function createLoggerFromLogging(
  logging: boolean | Logger | LogLevel | undefined,
) {
  if (logging == null || typeof logging === 'boolean') {
    return new Logger({ level: logging === false ? false : 'info' });
  }
  if (typeof logging === 'string') {
    return new Logger({ level: logging });
  }
  return logging;
}
