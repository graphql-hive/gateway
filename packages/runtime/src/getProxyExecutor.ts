import type {
  OnSubgraphExecuteHook,
  TransportEntry,
} from '@graphql-mesh/fusion-runtime';
import { getOnSubgraphExecute } from '@graphql-mesh/fusion-runtime';
import type { Executor } from '@graphql-tools/utils';
import type { GraphQLSchema } from 'graphql';
import type { GatewayConfigContext, GatewayConfigProxy } from './types';

export function getProxyExecutor<TContext extends Record<string, any>>({
  config,
  configContext,
  getSchema,
  onSubgraphExecuteHooks,
  transportExecutorStack,
}: {
  config: GatewayConfigProxy<TContext>;
  configContext: GatewayConfigContext;
  getSchema: () => GraphQLSchema;
  onSubgraphExecuteHooks: OnSubgraphExecuteHook[];
  transportExecutorStack: AsyncDisposableStack;
}): Executor {
  const fakeTransportEntryMap: Record<string, TransportEntry> = {};
  let subgraphName: string = 'upstream';
  const onSubgraphExecute = getOnSubgraphExecute({
    onSubgraphExecuteHooks,
    transportEntryMap: new Proxy(fakeTransportEntryMap, {
      get(fakeTransportEntryMap, subgraphNameProp: string): TransportEntry {
        if (!fakeTransportEntryMap[subgraphNameProp]) {
          subgraphName = subgraphNameProp;
          fakeTransportEntryMap[subgraphNameProp] = {
            kind: 'http',
            subgraph: subgraphName.toString(),
            location: config.proxy?.endpoint,
            headers: config.proxy?.headers as any,
            options: config.proxy,
          };
        }
        return fakeTransportEntryMap[subgraphNameProp];
      },
    }),
    transportContext: configContext,
    getSubgraphSchema: getSchema,
    transportExecutorStack,
    transports: config.transports,
  });
  return function proxyExecutor(executionRequest) {
    return onSubgraphExecute(subgraphName, executionRequest);
  };
}
