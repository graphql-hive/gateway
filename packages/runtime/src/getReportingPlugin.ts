import { useApolloUsageReport } from '@graphql-yoga/plugin-apollo-usage-report';
import useHiveConsole, {
  HiveConsolePluginOptions,
} from './plugins/useHiveConsole';
import type {
  GatewayConfig,
  GatewayConfigContext,
  GatewayPlugin,
} from './types';

export function getReportingPlugin<TContext extends Record<string, any>>(
  config: GatewayConfig<TContext>,
  configContext: GatewayConfigContext,
): {
  name?: 'Hive' | 'GraphOS';
  plugin: GatewayPlugin<TContext>;
} {
  if (config.reporting?.type === 'hive') {
    const { target, ...reporting } = config.reporting;
    let usage: HiveConsolePluginOptions['usage'] = reporting.usage;
    if (usage === false) {
      // explicitly disabled, leave disabled
    } else {
      // user specified a target, extend the usage with the given target
      usage = {
        target,
        ...(typeof usage === 'object' ? { ...usage } : {}),
      };
    }
    return {
      name: 'Hive',
      plugin: useHiveConsole({
        logger: configContext.logger.child({ reporting: 'Hive' }),
        enabled: true,
        ...reporting,
        ...(usage ? { usage } : {}),
        ...(config.persistedDocuments &&
        'type' in config.persistedDocuments &&
        config.persistedDocuments?.type === 'hive'
          ? {
              experimental__persistedDocuments: {
                cdn: {
                  endpoint: config.persistedDocuments.endpoint,
                  accessToken: config.persistedDocuments.token,
                },
                allowArbitraryDocuments:
                  !!config.persistedDocuments.allowArbitraryDocuments,
              },
            }
          : {}),
      }),
    };
  } else if (
    config.reporting?.type === 'graphos' ||
    (!config.reporting &&
      'supergraph' in config &&
      typeof config.supergraph === 'object' &&
      'type' in config.supergraph &&
      config.supergraph.type === 'graphos')
  ) {
    if (
      'supergraph' in config &&
      typeof config.supergraph === 'object' &&
      'type' in config.supergraph &&
      config.supergraph.type === 'graphos'
    ) {
      if (!config.reporting) {
        config.reporting = {
          type: 'graphos',
          apiKey: config.supergraph.apiKey,
          graphRef: config.supergraph.graphRef,
        };
      } else {
        config.reporting.apiKey ||= config.supergraph.apiKey;
        config.reporting.graphRef ||= config.supergraph.graphRef;
      }
    }

    return {
      name: 'GraphOS',
      // @ts-expect-error - TODO: Fix types
      plugin: useApolloUsageReport({
        agentVersion: `hive-gateway@${globalThis.__VERSION__}`,
        ...config.reporting,
      }),
    };
  }
  return {
    plugin: {},
  };
}
