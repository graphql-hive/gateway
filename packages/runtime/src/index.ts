export * from './createGatewayRuntime';
export { LogLevel, DefaultLogger } from '@graphql-mesh/utils';
export type * from './types';
export * from './plugins/useCustomFetch';
export * from './plugins/useStaticFiles';
export * from './getProxyExecutor';
export * from './plugins/usePropagateHeaders';
export * from '@whatwg-node/disposablestack';
export type { ResolveUserFn, ValidateUserFn } from '@envelop/generic-auth';
export * from '@graphql-mesh/hmac-upstream-signature';
export {
  getSdkRequesterForUnifiedGraph,
  getExecutorForUnifiedGraph,
} from '@graphql-mesh/fusion-runtime';
export { useUpstreamRetry } from './plugins/useUpstreamRetry';
export { useUpstreamTimeout } from './plugins/useUpstreamTimeout';
