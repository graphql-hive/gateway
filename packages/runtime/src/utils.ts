import type { SelectionSetNode } from 'graphql';

export function checkIfDataSatisfiesSelectionSet(
  selectionSet: SelectionSetNode,
  data: any,
): boolean {
  if (Array.isArray(data)) {
    return data.every((item) =>
      checkIfDataSatisfiesSelectionSet(selectionSet, item),
    );
  }
  for (const selection of selectionSet.selections) {
    if (selection.kind === 'Field') {
      const field = selection;
      const responseKey = field.alias?.value || field.name.value;
      if (data[responseKey] != null) {
        if (field.selectionSet) {
          if (
            !checkIfDataSatisfiesSelectionSet(
              field.selectionSet,
              data[field.name.value],
            )
          ) {
            return false;
          }
        }
      } else {
        return false;
      }
    } else if (selection.kind === 'InlineFragment') {
      const inlineFragment = selection;
      if (
        !checkIfDataSatisfiesSelectionSet(inlineFragment.selectionSet, data)
      ) {
        return false;
      }
    }
  }
  return true;
}

export const defaultQueryText = /* GraphQL */ `
  # Welcome to GraphiQL
  # GraphiQL is an in-browser tool for writing, validating,
  # and testing GraphQL queries.
  #
  # Type queries into this side of the screen, and you will
  # see intelligent typeaheads aware of the current GraphQL
  # type schema and live syntax and validation errors
  # highlighted within the text.
  #
  # GraphQL queries typically start with a "{" character.
  # Lines that start with a # are ignored.
  #
  # An example GraphQL query might look like:
  #
  #     {
  #       field(arg: "value") {
  #         subField
  #       }
  #     }
  #
`;
