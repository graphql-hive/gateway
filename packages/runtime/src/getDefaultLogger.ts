import { JSONLogger } from '@graphql-hive/logger-json';
import { process } from '@graphql-mesh/cross-helpers';
import { Logger } from '@graphql-mesh/types';
import { DefaultLogger, LogLevel } from '@graphql-mesh/utils';

export function getDefaultLogger(opts?: { name?: string; level?: LogLevel }) {
  const logFormat = process.env['LOG_FORMAT'] || (globalThis as any).LOG_FORMAT;
  if (logFormat) {
    if (logFormat.toLowerCase() === 'json') {
      return new JSONLogger(opts);
    } else if (logFormat.toLowerCase() === 'pretty') {
      return new DefaultLogger(opts?.name, opts?.level);
    }
  }
  const nodeEnv = process.env['NODE_ENV'] || (globalThis as any).NODE_ENV;
  if (nodeEnv === 'production') {
    return new JSONLogger(opts);
  }
  return new DefaultLogger(opts?.name, opts?.level);
}

export function handleLoggingConfig(
  loggingConfig:
    | boolean
    | Logger
    | LogLevel
    | keyof typeof LogLevel
    | undefined,
  existingLogger?: Logger,
) {
  if (typeof loggingConfig === 'object') {
    return loggingConfig;
  }
  if (typeof loggingConfig === 'boolean') {
    if (!loggingConfig) {
      if (existingLogger && 'logLevel' in existingLogger) {
        existingLogger.logLevel = LogLevel.silent;
        return existingLogger;
      }
      return getDefaultLogger({
        name: existingLogger?.name,
        level: LogLevel.silent,
      });
    }
  }
  if (typeof loggingConfig === 'number') {
    if (existingLogger && 'logLevel' in existingLogger) {
      existingLogger.logLevel = loggingConfig;
      return existingLogger;
    }
    return getDefaultLogger({
      name: existingLogger?.name,
      level: loggingConfig,
    });
  }
  if (typeof loggingConfig === 'string') {
    if (existingLogger && 'logLevel' in existingLogger) {
      existingLogger.logLevel = LogLevel[loggingConfig];
      return existingLogger;
    }
    return getDefaultLogger({
      name: existingLogger?.name,
      level: LogLevel[loggingConfig],
    });
  }
  return existingLogger || getDefaultLogger();
}
