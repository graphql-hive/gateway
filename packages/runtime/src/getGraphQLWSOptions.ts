// yoga's envelop may augment the `execute` and `subscribe` operations

import { MaybePromise } from '@graphql-tools/utils';
import { execute, subscribe, type ExecutionArgs } from 'graphql';
import type {
  ConnectionInitMessage,
  Context,
  ServerOptions,
  SubscribeMessage,
  SubscribePayload,
} from 'graphql-ws';
import type { GatewayRuntime } from './createGatewayRuntime';

// so we need to make sure we always use the freshest instance
type EnvelopedExecutionArgs = ExecutionArgs & {
  rootValue: {
    execute: typeof execute;
    subscribe: typeof subscribe;
  };
};

export function getGraphQLWSOptions<TContext extends Record<string, any>, E>(
  gwRuntime: GatewayRuntime<TContext>,
  onContext: (
    ctx: Context<ConnectionInitMessage['payload'], E>,
  ) => MaybePromise<Record<string, unknown>>,
): ServerOptions<ConnectionInitMessage['payload'], E> {
  return {
    execute: (args) => (args as EnvelopedExecutionArgs).rootValue.execute(args),
    subscribe: (args) =>
      (args as EnvelopedExecutionArgs).rootValue.subscribe(args),
    onSubscribe: async (ctx, idOrMessage, payloadOrUndefined) => {
      let payload: SubscribePayload;
      if (typeof idOrMessage === 'string') {
        // >=v6
        payload = payloadOrUndefined;
      } else {
        // <=v5
        payload = (idOrMessage as SubscribeMessage).payload;
      }

      const { schema, execute, subscribe, contextFactory, parse, validate } =
        gwRuntime.getEnveloped({
          connectionParams: ctx.connectionParams,
          ...(await onContext(ctx)),
        });
      const args: EnvelopedExecutionArgs = {
        schema: schema || (await gwRuntime.getSchema()),
        operationName: payload.operationName,
        document: parse(payload.query),
        variableValues: payload.variables,
        contextValue: await contextFactory(),
        rootValue: {
          execute,
          subscribe,
        },
      };
      if (args.schema) {
        const errors = validate(args.schema, args.document);
        if (errors.length) return errors;
      }
      return args;
    },
  };
}
