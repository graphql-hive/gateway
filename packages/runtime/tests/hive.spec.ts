import { useDisableIntrospection } from '@envelop/disable-introspection';
import { getUnifiedGraphGracefully } from '@graphql-mesh/fusion-composition';
import { printSchemaWithDirectives } from '@graphql-tools/utils';
import { createDisposableServer, isDebug } from '@internal/testing';
import { Response } from '@whatwg-node/fetch';
import { createServerAdapter } from '@whatwg-node/server';
import {
  buildClientSchema,
  getIntrospectionQuery,
  GraphQLSchema,
  printSchema,
  type ExecutionResult,
  type IntrospectionQuery,
} from 'graphql';
import { createSchema, createYoga } from 'graphql-yoga';
import { describe, expect, it, vi } from 'vitest';
import { createGatewayRuntime } from '../src/createGatewayRuntime';
import { useCustomFetch } from '../src/plugins/useCustomFetch';

function createUpstreamSchema() {
  return createSchema({
    typeDefs: /* GraphQL */ `
        """
        Fetched on ${new Date().toISOString()}
        """
        type Query {
          foo: String
        }
      `,
    resolvers: {
      Query: {
        foo: () => 'bar',
      },
    },
  });
}

describe('Hive CDN', () => {
  it('respects env vars', async () => {
    await using cdnServer = await createDisposableServer(
      createServerAdapter(
        () =>
          new Response(
            getUnifiedGraphGracefully([
              {
                name: 'upstream',
                schema: createUpstreamSchema(),
                url: 'http://upstream/graphql',
              },
            ]),
          ),
      ),
    );
    await using gateway = createGatewayRuntime({
      supergraph: {
        type: 'hive',
        endpoint: cdnServer.url,
        key: 'key',
      },
      logging: isDebug(),
    });
    const res = await gateway.fetch('http://localhost:4000/graphql', {
      method: 'POST',
      headers: {
        'content-type': 'application/json',
      },
      body: JSON.stringify({
        query: getIntrospectionQuery({
          descriptions: false,
        }),
      }),
    });

    expect(res.status).toBe(200);
    const resJson: ExecutionResult<IntrospectionQuery> = await res.json();
    const clientSchema = buildClientSchema(resJson.data!);
    expect(printSchema(clientSchema)).toMatchSnapshot('hive-cdn');

    // Landing page
    const landingPageRes = await gateway.fetch('http://localhost:4000', {
      method: 'GET',
      headers: {
        accept: 'text/html',
      },
    });
    const landingPage = await landingPageRes.text();
    expect(landingPage).toContain('Hive CDN');
    expect(landingPage).toContain('upstream');
    expect(landingPage).toContain('http://upstream/graphql');
  });
  it('uses Hive CDN instead of introspection for Proxy mode', async () => {
    const upstreamSchema = createUpstreamSchema();
    await using cdnServer = await createDisposableServer(
      createServerAdapter(() =>
        Response.json({
          sdl: printSchemaWithDirectives(upstreamSchema),
        }),
      ),
    );
    const upstreamServer = createYoga({
      schema: upstreamSchema,
      // Make sure introspection is not fetched from the service itself
      plugins: [useDisableIntrospection()],
    });
    let schemaChangeSpy = vi.fn((_schema: GraphQLSchema) => {});
    const hiveEndpoint = cdnServer.url;
    const hiveKey = 'key';
    await using gateway = createGatewayRuntime({
      proxy: { endpoint: 'http://upstream/graphql' },
      schema: {
        type: 'hive',
        endpoint: hiveEndpoint,
        key: hiveKey,
      },
      plugins: () => [
        useCustomFetch((url, opts): any => {
          if (url === 'http://upstream/graphql') {
            return upstreamServer.fetch(
              // @ts-expect-error TODO: url can be a string, not only an instance of URL
              url,
              opts,
            );
          }
          return gateway.fetchAPI.Response.error();
        }),
        {
          onSchemaChange({ schema }) {
            schemaChangeSpy(schema);
          },
        },
      ],
      logging: isDebug(),
    });
    const res = await gateway.fetch('http://localhost:4000/graphql', {
      method: 'POST',
      headers: {
        'content-type': 'application/json',
      },
      body: JSON.stringify({
        query: /* GraphQL */ `
          query {
            foo
          }
        `,
      }),
    });
    const resJson: ExecutionResult = await res.json();
    expect(resJson).toEqual({
      data: {
        foo: 'bar',
      },
    });
    expect(schemaChangeSpy).toHaveBeenCalledTimes(1);
    expect(printSchema(schemaChangeSpy.mock.calls[0]?.[0]!)).toBe(
      printSchema(upstreamSchema),
    );
  });
  it('handles persisted documents without reporting', async () => {
    const token = 'secret';
    await using cdnServer = await createDisposableServer(
      createServerAdapter((req) => {
        if (
          req.url.endsWith(
            '/apps/graphql-app/1.0.0/Eaca86e9999dce9b4f14c4ed969aca3258d22ed00',
          )
        ) {
          const hiveCdnKey = req.headers.get('x-hive-cdn-key');
          if (hiveCdnKey !== token) {
            return new Response('Unauthorized', { status: 401 });
          }
          return new Response(/* GraphQL */ `
            query MyTest {
              foo
            }
          `);
        }
        return new Response('Not Found', { status: 404 });
      }),
    );
    await using upstreamServer = await createDisposableServer(
      createYoga({
        schema: createSchema({
          typeDefs: /* GraphQL */ `
            type Query {
              foo: String
            }
          `,
          resolvers: {
            Query: {
              foo: () => 'bar',
            },
          },
        }),
      }),
    );
    await using gateway = createGatewayRuntime({
      proxy: {
        endpoint: `${upstreamServer.url}/graphql`,
      },
      persistedDocuments: {
        type: 'hive',
        endpoint: cdnServer.url,
        token,
      },
      logging: isDebug(),
    });
    const res = await gateway.fetch('http://localhost:4000/graphql', {
      method: 'POST',
      headers: {
        'content-type': 'application/json',
      },
      body: JSON.stringify({
        documentId:
          'graphql-app~1.0.0~Eaca86e9999dce9b4f14c4ed969aca3258d22ed00',
      }),
    });
    const resJson: ExecutionResult = await res.json();
    expect(resJson).toEqual({
      data: {
        foo: 'bar',
      },
    });
  });
  it('handles persisted documents with reporting', async () => {
    const token = 'secret';
    await using cdnServer = await createDisposableServer(
      createServerAdapter((req) => {
        if (
          req.url.endsWith(
            '/apps/graphql-app/1.0.0/Eaca86e9999dce9b4f14c4ed969aca3258d22ed00',
          )
        ) {
          const hiveCdnKey = req.headers.get('x-hive-cdn-key');
          if (hiveCdnKey !== token) {
            return new Response('Unauthorized', { status: 401 });
          }
          return new Response(/* GraphQL */ `
            query MyTest {
              foo
            }
          `);
        }
        return new Response('Not Found', { status: 404 });
      }),
    );
    await using upstreamServer = await createDisposableServer(
      createYoga<{}>({
        schema: createSchema({
          typeDefs: /* GraphQL */ `
            type Query {
              foo: String
            }
          `,
          resolvers: {
            Query: {
              foo: () => 'bar',
            },
          },
        }),
      }),
    );
    await using gateway = createGatewayRuntime({
      proxy: {
        endpoint: `${upstreamServer.url}/graphql`,
      },
      reporting: {
        type: 'hive',
        token: '',
      },
      persistedDocuments: {
        type: 'hive',
        endpoint: cdnServer.url,
        token,
      },
      logging: isDebug(),
    });
    const res = await gateway.fetch('http://localhost:4000/graphql', {
      method: 'POST',
      headers: {
        'content-type': 'application/json',
      },
      body: JSON.stringify({
        documentId:
          'graphql-app~1.0.0~Eaca86e9999dce9b4f14c4ed969aca3258d22ed00',
      }),
    });
    const resJson: ExecutionResult = await res.json();
    expect(resJson).toEqual({
      data: {
        foo: 'bar',
      },
    });
  });
});
