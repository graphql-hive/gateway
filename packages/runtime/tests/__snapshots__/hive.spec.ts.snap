// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Hive CDN respects env vars: hive-cdn 1`] = `
"type Query {
  foo: String
}"
`;
